### PR TITLE
Periodic Boundary Condition support

### PR DESCRIPTION
**Pull request by tbmasood**
Contribution:
- Added a class called PeriodicImplicitTriangulation which constructs triangulation of a grid with periodic boundary conditions.
- Modified the MorseSmaleComplex, FTMTree, TopologicalSimplification and PersistenceDiagram TTK Paraview plugins so that they have a checkbox to activate periodic boundary conditions to be used during computation.

### CODE REVIEW

File: core/base/PeriodicImplicitTriangulation/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,5 @@
+ttk_add_base_library(periodicImplicitTriangulation
```
Comment (julien-tierny) [inline core/base/PeriodicImplicitTriangulation/CMakeLists.txt]: to comply with ttk's naming convention, please rename (git mv)
- core/base/PeriodicImplicitTriangulation/ into  core/base/periodicImplicitTriangulation/
thanks for changing that in the involved cmake files.

File: core/base/triangulation/Triangulation.h [hunk 2]
```diff
@@ -3100,18 +3102,40 @@ namespace ttk {
...
       gridDimensions_[1] = yDim;
       gridDimensions_[2] = zDim;
 
-      return implicitTriangulation_.setInputGrid(xOrigin, yOrigin, zOrigin,
-                                                 xSpacing, ySpacing, zSpacing,
-                                                 xDim, yDim, zDim);
+      int ret = periodicImplicitTriangulation_.setInputGrid(
+        xOrigin, yOrigin, zOrigin, xSpacing, ySpacing, zSpacing, xDim, yDim,
+        zDim);
+      int retPeriodic = implicitTriangulation_.setInputGrid(
```
Comment (julien-tierny) [inline core/base/triangulation/Triangulation.h]: it seems that the variables "ret" and "retPeriodic" need to be switched.

File: core/base/PeriodicImplicitTriangulation/PeriodicImplicitTriangulation.h [hunk 3]
```diff
@@ -0,0 +1,13204 @@
+#ifndef _PERIODICIMPLICITTRIANGULATION_H
```
Comment (julien-tierny) [inline core/base/PeriodicImplicitTriangulation/PeriodicImplicitTriangulation.h]: stupid question: does **all** the API support periodicity? (or just a selection of the functions)

File: core/base/PeriodicImplicitTriangulation/CMakeLists.txt (hunk 1 above)
Comment (tbmasood) [reply core/base/PeriodicImplicitTriangulation/CMakeLists.txt]: Done.

File: core/base/PeriodicImplicitTriangulation/PeriodicImplicitTriangulation.h (hunk 3 above)
Comment (tbmasood) [reply core/base/PeriodicImplicitTriangulation/PeriodicImplicitTriangulation.h]: All functions are re-written to support periodicity.

File: core/base/triangulation/Triangulation.h (hunk 2 above)
Comment (tbmasood) [reply core/base/triangulation/Triangulation.h]: Corrected in latest commit.

**Comment by tbmasood** [comment]
In the latest commits, I have fixed the minor changes requested. Further, I have rewritten the code to make it more concise, reducing the lines of code to about a third of the original.

**Comment by julien-tierny** [comment]
alright, that looks ready to go.
cheers!
julien
-- 
Dr Julien Tierny
CNRS Researcher
Sorbonne Universite
http://lip6.fr/Julien.Tierny

**Pull request merged**